File: kotlin/circular-buffer/_iterations/1/src/main/kotlin/CircularBuffer.kt
class EmptyBufferException(message: String? = null, cause: Throwable? = null) : Exception(message, cause)
class BufferFullException(message: String? = null, cause: Throwable? = null) : Exception(message, cause)

class CircularBuffer<T>(capacity: Int) {
    private val data = MutableList<T?>(capacity) { null }
    private var size = 0
    private var p = 0
    private var g = 0

    val empty: Boolean
        get() = size == 0

    val full: Boolean
        get() = size == data.size

    fun read() : T {
        require(!empty) { EmptyBufferException("Buffer is empty") }

        val value = data[g]!!
        data[g] = null
        g = g.next
        --size
        return value
    }

    fun write(value: T) {
        require(!full) { BufferFullException("Buffer is full") }

        overwrite(value)
    }

    fun overwrite(value: T) {
        data[p] = value
        if (full) {
            g = g.next
        } else {
            ++size
        }
        p = p.next
    }

    fun clear() {
        data.replaceAll { null }
        size = 0
        p = 0
        g = 0
    }

    private fun require(condition: Boolean, exceptionProvider: () -> Exception) {
        if (!condition) {
            throw exceptionProvider()
        }
    }

    private val Int.next: Int
        get() = (this + 1) % data.size
}

File: ruby/circular-buffer/_iterations/1/circular_buffer.rb
class CircularBuffer
  class BufferEmptyException < StandardError; end
  class BufferFullException < StandardError; end

  def initialize(max_size)
    @max_size = max_size
    clear
  end

  def write(e)
    raise BufferFullException, "Buffer is full" if @size == @buffer.size
    write! e
  end

  def write!(e)
    @buffer[@p] = e
    @g = (@g + 1) % @buffer.size if @size == @buffer.size
    @p = (@p + 1) % @buffer.size
    @size += 1 if @size < @buffer.size
  end

  def read
    raise BufferEmptyException, "Buffer is empty" if @size == 0
    e = @buffer[@g]
    @g = (@g + 1) % @buffer.size
    @size -= 1
    e
  end

  def clear
    @buffer = Array.new(@max_size)
    @g = @p = @size = 0
  end
end
